### DISCUSSION TITLE
Python/Matplotlib

### DISCUSSION THREAD

**Discussion by nilskvilvang** [quarto-dev/quarto-cli#2606] [Q&A]
I just started looking at Quarto, thinking I might switch from RMarkdown (at some point). When running through the example code at the quarto web page, I can run the R and Julia examples. Still, when trying to run the python code chunk, I get this: "ModuleNotFoundError: No module named 'matplotlib'" - I have the latest Python (3.10). I run the latest RStudio (2022.07.1 Build 554).

**Comment by chrisvwx** [comment]
I guess you could try `python3 -m pip install matplotlib` though I am not a python guru at all, and this could be wrong.

I am getting a similar error to you (`ModuleNotFoundError: No module named 'nbclient'`) even though a pip command (`python3 -m pip install nbclient`) says that the module in question is already installed. I wonder if I somehow have a python version mismatch.

**Reply by rgaiacs** [reply]
Hi @christianpeel and thanks for the comment.

Quarto do it's best to locate your installation of Python. Can you run `quarto check` from your terminal and share the output?

The output that I have is

```
[✓] Checking Quarto installation......OK
      Version: 1.3.11
      Path: /opt/quarto/bin

[✓] Checking basic markdown render....OK

[✓] Checking Python 3 installation....OK
      Version: 3.10.4 (Conda)
      Path: /home/raniere/mambaforge/envs/jupyter/bin/python
      Jupyter: 4.9.2
      Kernels: conda-env-racehorse-py, conda-env-fastai-py, conda-env-exeter-py, conda-env-microradiography-py, conda-env-gan-py, conda-env-minimal-py, conda-env-frictionless-py, conda-env-yolov5-coco128-py, conda-env-first-carpal-py, conda-env-jupyter-py, conda-env-label-studio-converter-py, python3

[✓] Checking Jupyter engine render....OK

[✓] Checking R installation...........OK
      Version: 4.2.1
      Path: /usr/lib/R
      LibPaths:
        - /home/raniere/R/x86_64-pc-linux-gnu-library/4.2
        - /usr/local/lib/R/site-library
        - /usr/lib/R/site-library
        - /usr/lib/R/library
      rmarkdown: 2.16

[✓] Checking Knitr engine render......OK
```

It tells the where my Python installation is and this helps to know if there is a mismatch of Python instance.

**Reply by chrisvwx** [reply]
Thanks for the suggestion.   My output (below) looks similar to yours, though my focus is on Julia, and I don't use R.  To be clear, the basic Python-based Quarto "hello" test doc on the Quarto home page works for me, its just the Julia-based [doc](https://quarto.org/docs/computations/julia.html) that fails on me.

Here's the output of `quarto check`:
```
[✓] Checking Quarto installation......OK
      Version: 1.1.251
      Path: /Applications/quarto/bin

[✓] Checking basic markdown render....OK

[✓] Checking Python 3 installation....OK
      Version: 3.10.8
      Path: /usr/local/opt/python@3.10/bin/python3.10
      Jupyter: 5.0.0
      Kernels: julia--1.1, julia-1.7, julia-0.6, julia-1.5, julia-1.2, python3

[✓] Checking Jupyter engine render....OK

[✓] Checking R installation...........(None)

      Unable to locate an installed version of R.
      Install R from https://cloud.r-project.org/
```

**Reply by chrisvwx** [reply]
Just updated my Quarto version so that `quarto check` says that my Quarto version is `1.2.269`, with the same failure when trying to compile the basic Julia file.  I.e. with the [test Julia file](https://quarto.org/docs/computations/julia.html) in `jtest.qmd`, the command `quarto render jtest.qmd --to html` gives the following output:
```
Starting julia-1.7 kernel...Traceback (most recent call last):
  File "/Applications/quarto/share/jupyter/jupyter.py", line 21, in <module>
    from notebook import notebook_execute, RestartKernel
  File "/Applications/quarto/share/jupyter/notebook.py", line 17, in <module>
    from nbclient import NotebookClient
ModuleNotFoundError: No module named 'nbclient'
```

**Reply by rgaiacs** [reply]
@christianpeel Thanks for the extra information. From your output, you are using the Jupyter kernel named `julia-1.7`. You **must** install the `nbclient` in the `julia-1.7` kernel. Try

```
$ conda activate julia-1.7
$ conda install nbclient
```

**Reply by chrisvwx** [reply]
It looks like the only conda I have is in `~/.julia/conda` that was possibly installed as a part of IJulia. I see that there is a Julia package [Conda.jl](https://github.com/JuliaPy/Conda.jl) which was not immediately helpful. The Julia conda wanted to be initialized (`conda init bash`) after which `conda activate julia-1.7` says it can't find `julia-1.7`:
```
EnvironmentNameNotFound: Could not find conda environment: julia-1.7
You can list all discoverable environments with `conda info --envs`.
```
I'm not sure I believe that, since the command `quarto render jtest.qmd` now gives output that starts by saying it is 'starting the julia-1.7 kernel':
```
Starting julia-1.7 kernel...Traceback (most recent call last):
  File "/Applications/quarto/share/jupyter/jupyter.py", line 21, in <module>
    from notebook import notebook_execute, RestartKernel
  File "/Applications/quarto/share/jupyter/notebook.py", line 17, in <module>
    from nbclient import NotebookClient
  File "/Users/cpeel/.julia/conda/3/lib/python3.7/site-packages/nbclient/__init__.py", line 6, in <module>
    from .client import NotebookClient, execute  # noqa: F401
  File "/Users/cpeel/.julia/conda/3/lib/python3.7/site-packages/nbclient/client.py", line 18, in <module>
    from traitlets import (
ImportError: cannot import name 'Callable' from 'traitlets' (/Users/cpeel/.julia/conda/3/lib/python3.7/site-packages/traitlets/__init__.py)
```
 I don't know that this is a quarto problem, and it's definitely not a Matplotlib problem as described in this thread.  Maybe I should go to the Julia slack rather than having you solve Julia problems

**Reply by kdheepak** [reply]
I was able to work around this issue by installing `jupyter`, `nbclient` etc in the conda installation from Julia.

```julia
using Pkg;
Pkg.add("Conda")

using Conda
Conda.add("ipython") # probably not required
Conda.add("jupyter")
Conda.add("nbclient")
```
